python/pyarrow/_dataset.pyx

```diff
@@ -1403,6 +1403,25 @@ cdef class PartitioningFactory(_Weakrefable):
         return self.wrapped
 
 
+cdef vector[shared_ptr[CArray]] _partitioning_dictionaries(
+        Schema schema, dictionaries) except *:
+    cdef:
+        vector[shared_ptr[CArray]] c_dictionaries
+
+    dictionaries = list(dictionaries or [])[:len(schema)]
+    while len(dictionaries) < len(schema):
+        dictionaries.append(None)
+
+    for field, dictionary in zip(schema, dictionaries):
+        if (isinstance(field.type, pa.DictionaryType) and
+                dictionary is not None):
+            c_dictionaries.push_back(pyarrow_unwrap_array(dictionary))
+        else:
+            c_dictionaries.push_back(<shared_ptr[CArray]> nullptr)
```

Comment:
`dictionary` is ignored here? That doesn't sound right.

Reply:
I'll adopt the `Dict[str, Array]` pattern, which will remove this discrepancy

Reply:
The entry in `dictionaries` is still ignored when the field is not of dictionary type. Which is a user error of course, but we could maybe raise an exception in that case instead of silently ignoring it.